GH-9228: ZeroMQ - provide binding to ZeroMqMessageHandler

alessiomatricardi opened the pull request:
Closes #9228 

This PR provides the `bind` capability to the outbound channel adapter `ZeroMqMessageHandler`.

> Consider to implement that logic in the Lifecycle.start()

I tried, but the current implementation instantiate the socket within the class constructor, and some old tests relies on this mechanism.

alessiomatricardi committed:
first implementation

alessiomatricardi committed:
add docs

artembilan reviewed (changes requested):
Please, also add your name to the `@author` list of all the affected classes.
And verify changes locally like `./gradlew :spring-integration-zeromq:check`.

artembilan commented on spring-integration-zeromq/src/main/java/org/springframework/integration/zeromq/outbound/ZeroMqMessageHandler.java:
```diff
@@ -122,6 +151,26 @@ public ZeroMqMessageHandler(ZContext context, String connectUrl, SocketType sock
+				Mono.just(this.context.createSocket(this.socketType))
+						.publishOn(this.publisherScheduler)
+						.doOnNext((socket) -> this.socketConfigurer.accept(socket))
+						.doOnNext((socket) -> this.bindPort.set(bindSocket(socket, this.bindPort.get())))
```
Well, actually we do this in the `start()`: 
```
this.socketMonoSubscriber = this.socketMono.subscribe();
```
So, all good, however looking into this one more time I believe that we have to always renew this `socketMono` in the `start()`. After the `stop()` the environment might be changed, so or now way to connect or bind.
Therefore every time when we start this component we have to renew socket.


I'm also thinking about DRY principle here.
Looks like we now have two socket `Mono` which has difference only in these lines:
```
.doOnNext((socket) -> socket.connect(connectUrl.get()))
```
in opposed to:
```
.doOnNext((socket) -> this.bindPort.set(bindSocket(socket, this.bindPort.get())))
```
Essentially `if..else` according to the server/client preferences.
This could be just extracted into a `private` method where we would do `connect` or `bind`. 
And `monoSocket` would be just a single instance in the `start` without any duplication.

WDYT?

alessiomatricardi replied:
Completely agree with you. I was trying to touch less code as possible, but as you said this is the best approach

artembilan reviewed (changes requested):
Also this new functionality has to be mention in the `whats-new.adoc`. Like a new section in the end:
```
[[x6.4-zeromq-changes]]
=== ZeroMQ Changes

```

alessiomatricardi committed:
protected constructor in ZeroMqMessageHandlerSpec, expose them via ZeroMq

alessiomatricardi committed:
introduce ZeroMqUtils, for common Zero MQ utilities functions

alessiomatricardi committed:
use ZeroMqUtils.bindSocket in ZeroMqMessageProducer

alessiomatricardi committed:
refactor ZeroMqMessageHandler providing connectUrl and bindPort setters and simple constructors, following the same logic used for ZeroMqMessageProvider

alessiomatricardi committed:
fix tests to follow the new ZeroMqMessageHandler implementation

alessiomatricardi committed:
fix typo

alessiomatricardi committed:
add new updates to whats-new.adoc

alessiomatricardi commented on spring-integration-zeromq/src/main/java/org/springframework/integration/zeromq/dsl/ZeroMqMessageHandlerSpec.java:
```diff
@@ -52,6 +52,16 @@ protected ZeroMqMessageHandlerSpec(ZContext context, String connectUrl) {
+	 * Create an instance based on the provided {@link ZContext} and binding port.
+	 * @param context the {@link ZContext} to use for creating sockets.
+	 * @param port the port to bind ZeroMq socket to over TCP.
+	 * @since 6.2.6
```
Done!

alessiomatricardi commented on src/reference/antora/modules/ROOT/pages/zeromq.adoc:
```diff
@@ -146,7 +146,9 @@ ZeroMqMessageProducer zeroMqMessageProducer(ZContext context, MessageChannel out
 The `ZeroMqMessageHandler` is a `ReactiveMessageHandler` implementation to produce publish messages into a ZeroMQ socket.
 Only `SocketType.PAIR`, `SocketType.PUSH` and `SocketType.PUB` are supported.
-The `ZeroMqMessageHandler` only supports connecting the ZeroMQ socket; binding is not supported.
+This component can connect to the remote socket or bind onto TCP protocol with the provided or random port.
```
The logic now follows the one implemented in ZeroMqMessageProducer: if not provided, a random port will be chosen

alessiomatricardi commented:
Addressed almost every comment. I also added a typo in `ZeroMqHeaders`, but I've not added myself as author because no code has been touched. Let me know if I should do it

artembilan commented on spring-integration-zeromq/src/main/java/org/springframework/integration/zeromq/ZeroMqUtils.java:
```diff
@@ -0,0 +1,54 @@
+	 */
+	public static int bindSocket(ZMQ.Socket socket, int port) {
+		if (port == 0) {
+			return socket.bindToRandomPort("tcp://*");
```
I wonder if it always has to be `tcp` protocol: http://api.zeromq.org/4-2:zmq-bind.

Or at least, let's document that in this method Javadoc!

alessiomatricardi replied:
Fully agree with you.
I only merged the behaviour between `ZeroMqMessageProducer` and `ZeroMqMessageHandler`.

I believe that the support of the other transport protocols is out of the scope of this PR, if you don't mind I will make a specific one for this purpose.
Additionally, some protocols like `inproc` don't need a port but only an arbitrary string endpoint, so some rework is needed.

BTW, I will remark the fact that bind is actually possible only over TCP.

artembilan replied:
Yeah... Let's look into that when concern arises!

artembilan commented on spring-integration-zeromq/src/main/java/org/springframework/integration/zeromq/outbound/ZeroMqMessageHandler.java:
```diff
@@ -206,6 +246,47 @@ public void wrapTopic(boolean wrapTopic) {
+	 * @param port the port to bind ZeroMq socket to over TCP.
+	 * @since 6.4
+	 */
+	public void setBindPort(int port) {
```
We have these options covered by constructors.
I don't see a reason in extra setters on the matter: might cause some paradox of choice.

artembilan commented on spring-integration-zeromq/src/main/java/org/springframework/integration/zeromq/outbound/ZeroMqMessageHandler.java:
```diff
@@ -91,6 +94,27 @@ public class ZeroMqMessageHandler extends AbstractReactiveMessageHandler
+		this(context, SocketType.PAIR);
+	}
+
+	public ZeroMqMessageHandler(ZContext context, SocketType socketType) {
```
These two new constructors need some Javadoc.
I guess they are about random port.

artembilan commented on spring-integration-zeromq/src/main/java/org/springframework/integration/zeromq/outbound/ZeroMqMessageHandler.java:
```diff
@@ -214,6 +295,8 @@ public String getComponentType() {
 	protected void onInit() {
 		super.onInit();
+		Assert.state(this.connectUrl == null || this.bindPort.get() == 0,
+				"Only one of the 'connectUrl' or 'bindPort' must be provided");
```
If we remove those new setters, this condition won't be necessary.
So, you would only have a choice to give `connectUrl` or `port` via those constructors.

artembilan commented on spring-integration-zeromq/src/main/java/org/springframework/integration/zeromq/dsl/ZeroMq.java:
```diff
@@ -58,6 +59,17 @@ public static ZeroMqMessageHandlerSpec outboundChannelAdapter(ZContext context,
+	 * @return the spec.
+	 * @since 6.4
+	 */
+	public static ZeroMqMessageHandlerSpec outboundChannelAdapter(ZContext context, int port) {
```
I wonder if DSL has to expose extra factories to explicitly bind to random port.
Something similar what you have with those new ctors in the `ZeroMqMessageHandler`.

artembilan replied:
Or you have missed to push the change, or just didn't implement yet.
Hence the conversion is not resolved yet 😄

artembilan reviewed (changes requested):
This is great!
Just couple follow up suggestions.

Thank you!

alessiomatricardi committed:
address ZeroMQUtils comments

alessiomatricardi committed:
remove connectUrl and boundPort setters, add Javadoc to new constructors

alessiomatricardi committed:
add new DLS constructor for random port

alessiomatricardi commented:
Addressed your comments!

artembilan commented on spring-integration-zeromq/src/main/java/org/springframework/integration/zeromq/ZeroMqUtils.java:
```diff
@@ -0,0 +1,51 @@
+ * Module that wraps common methods of ZeroMq integration classes
+ *
+ * @author Alessio Matricardi
+ *
```
Sorry, I might was not right: the `@since` is required for new classes. And after that for new `public` methods added to existing classes.
So, bring back `@since 6.4` over here, please

artembilan commented on spring-integration-zeromq/src/main/java/org/springframework/integration/zeromq/dsl/ZeroMqMessageHandlerSpec.java:
```diff
@@ -52,6 +52,25 @@ protected ZeroMqMessageHandlerSpec(ZContext context, String connectUrl) {
 	}
 
+	/**
+	 * Create an instance based on the provided {@link ZContext}.
```
Probably better to say that socket is going to be bound to random port.

artembilan commented on spring-integration-zeromq/src/main/java/org/springframework/integration/zeromq/dsl/ZeroMqMessageHandlerSpec.java:
```diff
@@ -73,6 +92,29 @@ protected ZeroMqMessageHandlerSpec(ZContext context, String connectUrl, SocketTy
 	}
 
+	/**
+	 * Create an instance based on the provided {@link ZContext} and {@link SocketType}.
```
DITTO

artembilan reviewed (changes requested):
Nice progress!
I believe we are very close for merge.
Please, take a look to the latest review.
Honestly, nothing critical and I can address them on merge.
Just let me know your preferences!

alessiomatricardi committed:
add since closure in ZeroMqUtils

alessiomatricardi committed:
add DSL support methods and specific that, when not defined, the socket will be bound to a random port

alessiomatricardi commented:
Hi, sorry for the delay but I'm very busy with work.. hope now looks good for you :)

artembilan commented:
@alessiomatricardi ,

thank you very much for the contribution; looking forward for more!
